### DISCUSSION TITLE
How to free memory of a variable when the variable is used in a function

### DISCUSSION THREAD

**Discussion by ghost** [edubart/nelua-lang#137] [Q&A] [answered]
Let's say,

I disabled the GC, so now i have to free the dynamically allocated memory.

But i want to free the memory only when the program is exited because that's how my program works.

How would i know, that the program is exited so now i can free up the memory.

**Comment by codehz** [comment]
You don't need free any memory/close file handle when program exit, the OS will do it for you
The only negative effect is that it will be detected by some tools like valgrind, but it is harmless...

**Reply by ghost** [reply]
even if the GC is disabled?

**Reply by codehz** [reply]
yes

**Comment by ghost** [comment]
but is it preferred to do it that way?

according to some guy on the internet :
> Because with that mindset when you working on a big project you will definitely cause memory leaks.
Modern OS will do it for you, who knows what an embedded system or an old version of an OS will do.

is there any other good solution?

**Reply by codehz** [reply]
You need to understand the memory leak before solving it.
Usually memory leaks refer to the increasing memory usage during execution. If you decide to release the memory only when the program exits, then this operation must be able to be eliminated without causing actual memory leaks (it is true for one-time program like compiler).
If you really care about memory leaks, you won't adopt this strategy. Usually you should release memory when it is no longer needed.
Anyway, you cannot do anything for this process **after** the process died..

When it comes to embedded platforms, if you say that it cannot reclaim memory before the program exits, it actually means that there is no concept of process at all in this platform. In fact, if you have really tried developing in such a platform, you will know that dynamic memory itself is a lie, but you decide which addresses to use for what purpose (you are using all the available memary all the time, no alloc, no free, you could implement a custom memory allocator for it, but you will quickly realize that there is nothing worth recycling at the end of the program, the only role of the allocator is to reuse memory during execution)

**Comment by ghost** [comment]
> If you really care about memory leaks, you won't adopt this strategy. Usually you should release memory when it is no longer needed.

That's why the topic of this discussion is "How to check if a nelua program exited with code 0"
Though it would make more sense if it would have been like this "I want to free memory when the program exists" ( sorry for that )

Let's say i have a string which is being used by a function in ``function.nelua``,  and that function is called from a file called as ``main.nelua``
```lua
-- function.nelua
## pragmas.nogc = true
require "string"

local mystring= string.create(5)
mystring= "\x1B[31m"

global function test()
      print(mystring)
end
```
```lua
-- main.nelua
test()
```

I want to free the memory of the variable ``mystring`` after the function is called in the ``main.nelua`` <br>

Now because i'have been thinking about it while writing this..
the possible solution might be, calling ``mystring:destroy`` in the end of the ``function.nelua`` file?

If that's the case then i have wasted codehz's time

**Comment by codehz** [comment, marked as answer]
you could use `defer` to make it called before function end 
sample:
```lua
local str = string.create("test")
defer str:destroy() end
# whatever
```
another method: use __close( search it in https://nelua.io/overview/ )
example:
```lua
local str <close> = string.create("test")
# ...
```
warn: if you want to return the owned string, take care about it, you cannot destroy it before return, and you don't want to leak it if exception was throw (aka forgot to return the string in middle)

**Comment by edubart** [comment]
@codehz said valid points, and you could just disable the GC at compile-time and never deallocate. But I would also add that another option to avoid leaks when using valgrind like tools, is to disable the GC at runtime instead of compile-time, calling `collectgarbage('stop')` as the first thing when the application starts, this way the GC will only collect garbage before the application finishes, in addition you could manually collect garbage in the middle of the program if needed.

**Reply by ghost** [reply]
Okay, didn't know that GC can be disabled for runtime and compile time separately

but now i do, thanks :+1:
